state/storage_test.go

```diff
@@ -844,6 +851,100 @@ func (s *StorageStateSuite) TestAttachStorageAssignedMachineExistingVolumeAttach
+
+	// Detach, but do not destroy, the storage.
+	err := s.State.DetachStorage(storageTag, u.UnitTag())
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.State.RemoveStorageAttachment(storageTag, u.UnitTag())
+	c.Assert(err, jc.ErrorIsNil)
+
+	ch, _, err := app.Charm()
+	c.Assert(err, jc.ErrorIsNil)
+	app2, err := s.State.AddApplication(state.AddApplicationArgs{
+		Name:   "secondwind",
+		Series: app.Series(),
+		Charm:  ch,
+		Storage: map[string]state.StorageConstraints{
+			// The unit should have two storage instances
+			// in total. We're attaching one, so only one
+			// new instance should be created.
+			"data": makeStorageCons("modelscoped", 1024, 2),
+		},
+		AttachStorage: []names.StorageTag{storageTag},
+		NumUnits:      1,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	app2Units, err := app2.AllUnits()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(app2Units, gc.HasLen, 1)
+
+	// Now attach the storage to the second unit.
+	storageInstance, err := s.State.StorageInstance(storageTag)
+	c.Assert(err, jc.ErrorIsNil)
+	owner, hasOwner := storageInstance.Owner()
+	c.Assert(hasOwner, jc.IsTrue)
+	c.Assert(owner, gc.Equals, app2Units[0].UnitTag())
+	storageAttachments, err := s.State.UnitStorageAttachments(app2Units[0].UnitTag())
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(storageAttachments, gc.HasLen, 2)
+}
+
+func (s *StorageStateSuite) TestAddApplicationAttachStorageMultipleUnits(c *gc.C) {
+	app, _, storageTag := s.setupSingleStorageDetachable(c, "block", "modelscoped")
+	ch, _, _ := app.Charm()
+	_, err := s.State.AddApplication(state.AddApplicationArgs{
+		Name:          "secondwind",
+		Series:        app.Series(),
+		Charm:         ch,
+		AttachStorage: []names.StorageTag{storageTag},
+		NumUnits:      2,
+	})
+	c.Assert(err, gc.ErrorMatches, `cannot add application "secondwind": AttachStorage is non-empty, but NumUnits is 2`)
+}
+
+func (s *StorageStateSuite) TestAddApplicationAttachStorageTooMany(c *gc.C) {
+	app, _, _ := s.setupSingleStorageDetachable(c, "block", "modelscoped")
+
+	// Create 3 units whose storage instances we'll detach,
+	// in order to attach to the new application below. The
+	// charm allows a maximum of 2 storage instances, so the
+	// application creation should fail.
+	var storageTags []names.StorageTag
+	for i := 0; i < 3; i++ {
+		u, err := app.AddUnit()
+		c.Assert(err, jc.ErrorIsNil)
+		storageTag := names.NewStorageTag("data/" + fmt.Sprint(i+1))
+		storageTags = append(storageTags, storageTag)
+
+		// Detach, but do not destroy, the storage.
+		err = s.State.DetachStorage(storageTag, u.UnitTag())
+		c.Assert(err, jc.ErrorIsNil)
+		err = s.State.RemoveStorageAttachment(storageTag, u.UnitTag())
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	ch, _, err := app.Charm()
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = s.State.AddApplication(state.AddApplicationArgs{
+		Name:   "secondwind",
+		Series: app.Series(),
+		Charm:  ch,
+		Storage: map[string]state.StorageConstraints{
```

Comment:
Storage defines how to *create* storage for new units; each storage constraint contains a Count. We'll create Count less the number of storage instances in AttachStorage (for the corresponding charm storage name) for the first unit, and the full Count for subsequent units.

e.g. if I do `juju deploy ceph-osd --storage osd-devices=3,100G --attach-storage osd-devices/0`, we'll create and attach two new osd-devices storage instances, and attach the existing osd-devices/0, to the first unit. If I then `juju add-unit ceph-osd`, three new storage instances will be created and attached to the new unit.